RFC / WIP: no longer warn in code_warntype for variables not used in the body

KristofferC opened the pull request:
The fact that `@code_warntype` shows unused variables in red is something that keeps tripping up users (both new and experienced), see e.g. https://discourse.julialang.org/t/question-about-type-stability-of-arrays-in-julia-0-6/5389.

This PR traverses the codeinfo to check if the variable is actually used somewhere in the body. If not, we no longer print the variable in red even if it is inferred as a non concrete type. 

Before:

![image](https://user-images.githubusercontent.com/1282691/29356286-0ebfa9a8-8274-11e7-8e12-f910d14f8aa3.png)

After:

![image](https://user-images.githubusercontent.com/1282691/29356292-13aeb90e-8274-11e7-9821-047e7e6356f4.png)


Maybe the `(Unused in body)` is too verbose?

Will add test / docs later if people think this is a good idea. Also, I'm new to the codeinfo data structure so someone should probably look over that code. CI skipped as to not waste resources so don't merge.

yuyichao commented on base/interactiveutil.jl:
```diff
@@ -347,14 +347,35 @@ problematic for performance, so the results need to be used judiciously.
+        return false
+    end
+
+    is_variable_used(arg::SlotNumber, slotnames, v) = slotnames[arg.id] == v
```
Compare the id directly.

yuyichao replied:
Also should be `::Slot`instead.

yuyichao commented on base/interactiveutil.jl:
```diff
@@ -347,14 +347,35 @@ problematic for performance, so the results need to be used judiciously.
 See [`@code_warntype`](@ref man-code-warntype) for more information.
 """
 function code_warntype(io::IO, f, @nospecialize(t))
+    function is_variable_used(bodies::Vector, slotnames, v)
```
This is also wrong. This can incorrectly catch constant arrays spliced into the code.

oscardssmith commented:
Would it make sense to make things yellow if unused in body? I feel like that would be a useful piece of info (although code warntype might not be a good place to show it)

yuyichao commented:
What's this info (unused vars) useful for? It's just a signature of the optimization.

It might also be better to just do a single scan of slots uses instead of multiple ones and there's no need to use dispatch for that.

KristofferC commented:
I changed to only doing a single scan. I am not sure how I would avoid using dispatch and if the Vector dispatch function is still problematic.

yuyichao commented:
```
    function scan_exprs(exprs, used)
        for expr in exprs
            if isa(ex, Slot)
                used[ex.id] = true
            elseif isa(ex, Expr)
                scan_expr(ex, used)
            end
        end
    end
    function slot_usaged(ci, slotnames)
        used = falses(length(slotnames))
        scan_exprs(ex, used)
        return used
    end
```

yuyichao commented on base/interactiveutil.jl:
```diff
@@ -346,15 +346,33 @@ This serves as a warning of potential type instability. Not all non-leaf types a
 See [`@code_warntype`](@ref man-code-warntype) for more information.
 """
-function code_warntype(io::IO, f, @nospecialize(t))
+function code_warntype(io::IO, f, t::ANY) #@nospecialize(t))
```
Also, I imagine this is for testing only but this should be changed back....

KristofferC replied:
Yeah

yuyichao commented:
Note that the code above rely on the AST to be valid. You can add `ex.id <= length(used) && (used[ex.id] = true)` to guard it if you want.

KristofferC committed:
no longer warn in code_warntype for unused variables

KristofferC commented:
Thanks! Updated with that code. Also added a test.

Now for the details. I currently avoid printing the inferred type and just print `(unused)` when the variable is not used. Do people like that or are there other suggestions for other ways of printing that gives the same information (not used + inferred type is not important for performance)?

KristofferC commented:
With some inspiration from GDB I went with:

![image](https://user-images.githubusercontent.com/1282691/29486316-282df46e-84e3-11e7-9c27-b7e306af782c.png)

Any opinions on that?

KristofferC commented:
Not sure if eligible for backport but tentatively putting the label there.

KristofferC commented:
Will merge tomorrow unless [the usual stuff].

fredrikekre commented:
Could we do something like this for the red `Union{}` on `throw` lines as well?
![screenshot from 2017-08-22 16-25-47](https://user-images.githubusercontent.com/11698744/29570385-a117dda2-8756-11e7-9ddf-fb9cdf346a2e.png)

KristofferC commented:
Yeah, should be fairly easy.
